app/controllers/users_controller.rb

```diff
@@ -1589,6 +1603,64 @@ def register_second_factor_security_key
     render json: failed_json.merge(error: err.message)
   end
 
+  def create_passkey
+    raise Discourse::NotFound unless SiteSetting.experimental_passkeys
+
+    challenge_session = DiscourseWebauthn.stage_challenge(current_user, secure_session)
+    render json:
+             success_json.merge(
+               challenge: challenge_session.challenge,
+               rp_id: DiscourseWebauthn.rp_id,
+               rp_name: DiscourseWebauthn.rp_name,
+               supported_algorithms: ::DiscourseWebauthn::SUPPORTED_ALGORITHMS,
+               user_secure_id: current_user.create_or_fetch_secure_identifier,
+               existing_passkey_credential_ids: current_user.passkey_credential_ids,
```

Comment:
Just curious here, when creating a new passkey, why do we need to pass down the credential ids of the other security keys?

Reply:
Also this returns `Securitykey` where `factory_type` is the first factor. Only passkeys are first factor for now but will it become a problem in the future if there are other types of `UserSecurityKey` which are first factor too?

Reply:
For the first question: we pass the credential ids here so that the frontend can pass them to the browser (or OS) in an `excludeCredentials` parameter. Most browsers will refuse to create a new passkey if `excludeCredentials` contains the identifier of a key that the browser already has. 

When I try locally in Chrome to add a second passkey, for example, I see this: 

<img width="475" alt="image" src="https://github.com/discourse/discourse/assets/368961/6e0e9f68-7bcc-4382-9724-c1c08432d8b4">

> Only passkeys are first factor for now but will it become a problem in the future if there are other types of UserSecurityKey which are first factor too?

A first-factor key is equal to a passkey. I don't think there can be a `UserSecurityKey` that is `first-factor` but not a passkey.

Reply:
Thank you for clarifying :+1: